fix: do not fail tracer for malformed uri

mpmiszczyk opened the pull request:
Failing tracer will be detached which can brake tracing all together.

mpmiszczyk committed:
fix: do not fail tracer for malformed uri

Failing tracer will be detached which can brake tracing all together.

mpmiszczyk committed:
Merge remote-tracking branch 'origin/master' into fix-failing-url-parsing-fresha

kamilkowalski commented on lib/spandex_phoenix.ex:
```diff
@@ -260,11 +260,17 @@ defmodule SpandexPhoenix do
+  defp uri_decode(path_component) do
+    URI.decode(path_component)
+  rescue
+    _error -> "<malformed_uri>"
```
I'm wondering if we couldn't just use the original request path - we're losing information that could be used elsewhere. If anyone wants to filter malformed URIs because their observability backend doesn't support them, they can always filter metadata using `customize_metadata` and use the output of `SpandexPhoenix.default_metadata/1`.

mpmiszczyk replied:
Few things (and sorry if I'm being too blunt in some parts - I'm trying to make sure that I'm understanding this correctly):

1. Using the original request path - that's a good point.  I thought that is our setup we could always pick up the original path from the Nginx log, but that would not work for everyone (and would put too much requirements on our logs setup).  In the end I wanted to reduce the number of resources in DD, but for those are responsible two different parts of the `%Spandex.Trace`.  So currently the `url:` will consist of full path, and the `resource:` will replace malformed parts with `<malformed>`.  And I've added log, just to make those errors more visible.
2. The resource name - current implementation is still not ideal, but I think it might be the most practical.  At the begging I thought that all those not parable uri should have one single resource `/<malformed>`, but this turns out to be tricky. It would require introducing some additional complexity, and with that some overhead. Leaving it as it is will produce resources which are mix of parsable and not parsable parts; something [like `GET /multiple/<malformed_uri>`](https://github.com/surgeventures/spandex_phoenix/pull/4/files#diff-32563a9c54d7d2adf1cd06f95b70cce481abdfbd6d3f1e134171044010b1b046R169).  This should be good enough, and without any additional overhead.
3. The detached tracer - I think I'm already doing that.  The `customize_metadata` is being called in the tracer process, and if it fails for any reason, the test process will not receive the processed trace.  So the test will fail [on the `assert_receive`](https://github.com/surgeventures/spandex_phoenix/pull/4/files#diff-32563a9c54d7d2adf1cd06f95b70cce481abdfbd6d3f1e134171044010b1b046R161) macro. Something like:

```elixir
     Assertion failed, no matching message after 100ms
     The process mailbox is empty.
     code: assert_receive {:sent_trace, %Spandex.Trace{spans: [%Spandex.Span{error: error, http: http, name: "request", resource: "GET /multiple/<malformed_uri>"}]}}
     stacktrace:
       test/tracer_integration/plug/router_test.exs:161: (test)
```

This still does not guarantee that all traces will succeed, or that the tracer will never be detached, but is should cover the case we discovered.  I've spend some time trying to find out if there are any other places where I could write any tests for this change, but came out empty.

kamilkowalski replied:
Point 1: I mostly agree, minus error logging (see my code review comment).
Point 2: Fully agree.
Point 3: If you mean my comment about us not testing telemetry handlers - what you're testing is the Router integration, but we're using the Telemetry integration (via `SpandexPhoenix.Telemetry.install/1`) and that's not tested at all.

kamilkowalski reviewed (commented):
The issue we encountered is when the `SpandexPhoenix.default_metadata/1` function was used in a telemetry handler, it would throw an error and de-register the handler - can we add a test case for that which verifies if the handler stays registered?

mpmiszczyk committed:
fix: only the resource is overridden with <malformed_uri>

mpmiszczyk committed:
Log errors on trace parsing

This still produces not ideal resource names, but it's hard to say
what would be and ideal one.  Both this way of parsing, and replacing
whole resource name with `"<malformed_uri>"` has it's own benefits.

mpmiszczyk committed:
chore: update version

mpmiszczyk committed:
Merge branch 'master' of github.com:surgeventures/spandex_phoenix into fix-failing-url-parsing-fresha

mpmiszczyk committed:
chor: new version for release

mpmiszczyk committed:
fix: handle unexpected errors in traces

Even if the processing of trace will fail for some unknown reason, the
attached handler should not pass up - this will cause detachment of
the handler.

And since we are not using any state here (other than minimal state
stored in `conn`, which itself is unique for each request), errors in
single trace do not have impact on other traces, hence no cleanup
needed.

kamilkowalski commented on lib/spandex_phoenix.ex:
```diff
@@ -260,11 +262,22 @@ defmodule SpandexPhoenix do
-    decoded_component = URI.decode(path_component)
+    case uri_decode(path_component) do
+      nil ->
+        Logger.error("malformed uri", malformed_uri: path_component)
```
This could easily flood logs in case of a DoS attack and it's not really an application error - it's the client sending an invalid payload. I don't think we should log it.

kamilkowalski commented on lib/spandex_phoenix/telemetry.ex:
```diff
@@ -111,6 +113,13 @@ defmodule SpandexPhoenix.Telemetry do
+        crash_reason: {error, __STACKTRACE__},
+        event: event,
+        conn: conn
+      )
```
This could also spam us easily if there's a route that always throws an error. I think the telemetry handler de-registration in case of exceptions is there for a reason - we should monitor it and fix the issue if it happens, and if it's the request that's wrong (client error) we should sit quietly.

mpmiszczyk committed:
Remove logs that might flood the system.

kamilkowalski reviewed (approved):
:100:
